kaos/algorithm/interpolator.py

```diff
@@ -0,0 +1,114 @@
+            segments include this timestamp, which may occur if timestamp is a segment
+            boundary, then return the segment whose start time is timestamp. If no matching
+            segment exists, return None.
+        """
+        # TODO move this to model
+        try:
+            # pylint: disable=undefined-variable
+            return (OrbitSegments.query.filter(OrbitSegments.platform_id == platform_id,
+                                               OrbitSegments.start_time <= timestamp,
+                                               OrbitSegments.end_time >= timestamp)
+                                        .order_by(OrbitSegments.start_time.desc())
+                                        .first())
+
+            # pylint: enable=undefined-variable
+        except NoResultFound:
+            return None
+
+    @staticmethod
+    def get_orbit_records_by_segment(segment_id):
+        """Get all the orbit records in a segment.
+
+        Args:
+            segment_id: The segment id.
+
+        Return:
+            All the orbit records, sorted by time.
+        """
+        # TODO move this to model
+        try:
+            # pylint: disable=undefined-variable
+            orbit_records = (OrbitRecords.query.filter_by(segment_id=segment_id)
+                                               .order_by(OrbitRecords.time))
+            # pylint: enable=undefined-variable
+            return list(orbit_records)
+        except NoResultFound:
+            return []
+        
+    @staticmethod
+    def linear_interp(platform_id, timestamp):
+        """Use linear interpolation to estimate the position and velocity of a satellite 
+        at a given time.
+
+        Args:
+            platform_id: The UID of the satellite.
+            timestamp: The time for which to get the estimated position and velocity.
+
+        Return:
+            A tuple (pos, vel). Both pos, vel are 3-tuples containing the vector components
+            of the position and velocity. Return None if interpolation could not be done
+            for the given satellite and timestamp.
+        """
+        # find the correct segment
+        segment = Interpolator.get_segment(platform_id, timestamp)
+        if not segment:
+            return None
+
+        orbit_records = Interpolator.get_orbit_records_by_segment(segment.segment_id)
+
+        # can't do linear interpolation in this case
+        if not orbit_records or len(orbit_records) < 2:
+            return None
+        
+        xp = np.array([rec.time for rec in orbit_records])
+        pos = np.array([np.array(rec.position) for rec in orbit_records])
+        vel = np.array([np.array(rec.velocity) for rec in orbit_records])
+
+        # TODO there's probably an elegant 1-liner way to do this with numpy
+        tpos = np.zeros(3)
+        tvel = np.zeros(3)
+
+        for i in range(3):
+            tpos[i] = np.interp(x=timestamp, xp=xp, fp=pos[:, i])
```

Comment:
Wait what, your xp and fp have different dimensions, is that ok?

Reply:
They have the same dimensions. That slicing notation takes a vertical slice.

Reply:
Oh right its a list of lists, derp.